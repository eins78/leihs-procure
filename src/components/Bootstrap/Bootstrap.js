import React, { Fragment as F } from 'react'
import PropTypes from 'prop-types'
import cx from 'classnames'
import f from 'lodash'

import BsButton from 'reactstrap/lib/Button'

import Icon from '../Icons'
import { ControlledInput } from './ControlledForm'

const BOOTSTRAP_BREAKPOINTS = ['sm', 'md', 'lg', 'xl']
const BOOTSTRAP_MODIFIERS = [
  'primary',
  'secondary',
  'success',
  'danger',
  'warning',
  'info',
  'light',
  'dark'
]

// https://getbootstrap.com/docs/4.0/utilities/spacing/#notation
const bsSizeUtils = ({ m, p, ...props }) => {
  const { cls, className, ...restProps } = props
  function extract(letter, prop) {
    if (f.isArray(prop)) return prop.map(p => extract(letter, p))
    if (!f.isString(prop)) return
    let parts = prop.split(' ')
    if (parts.length > 1) return extract(letter, parts)
    return `${letter}${prop}`
  }
  const margin = extract('m', m)
  const padding = extract('p', p)
  return [cx(margin, padding, cls, className), restProps]
}

const Node = ({ tag = 'div', ...props }) => {
  // eslint-disable-next-line no-debugger
  const Tag = tag
  const [bsClasses, restProps] = bsSizeUtils(props)
  return <Tag {...restProps} className={bsClasses} />
}
Node.propTypes = {
  tag: PropTypes.string
}

export { Node as __Node }

export const Div = props => Node(props)
export const Pre = props => Node({ ...props, tag: 'span' })
export const Span = props => Node({ ...props, tag: 'span' })
export const Code = props => Node({ ...props, tag: 'span' })
export const Hr = props => Node({ ...props, tag: 'span' })
export const Main = props => Node({ ...props, tag: 'main' })
export const Label = props => Node({ ...props, tag: 'label' })
export const Small = props => Node({ ...props, tag: 'small' })

export const Container = ({ fluid, ...props }) =>
  Node({ ...props, cls: [fluid ? 'container-fluid' : 'container', props.cls] })

export const Row = ({ form, ...props }) =>
  Node({ ...props, cls: [form ? 'form-row' : 'row', props.cls] })

export const Col = ({ order, cls, ...props }) => {
  const restProps = f.omit(props, BOOTSTRAP_BREAKPOINTS)
  const breakpoint = f.first(
    f.intersection(f.keys(props), BOOTSTRAP_BREAKPOINTS)
  )
  const breakpointVal = props[breakpoint]
  const colCls = breakpoint
    ? f.isString(breakpointVal) || f.isNumber(breakpointVal)
      ? `col-${breakpoint}-${breakpointVal}`
      : `col-${breakpoint}`
    : 'col'
  const orderCls =
    order && (breakpoint ? `order-${breakpoint}-${order}` : `order-${order}`)
  return Node({ ...restProps, cls: cx(colCls, orderCls, cls) })
}

export const Button = ({ flat, className, cls, ...props }) => (
  <BsButton
    type="button" // default in case not given, otherwise its 'submit'
    {...props}
    className={cx(className, cls, { 'btn-flat': flat })}
  />
)
Button.propTypes = {
  className: PropTypes.string,
  flat: PropTypes.bool
}

export const Badge = props => {
  const restProps = f.omit(props, BOOTSTRAP_MODIFIERS)
  const mod =
    f.first(f.intersection(f.keys(props), BOOTSTRAP_MODIFIERS)) ||
    BOOTSTRAP_MODIFIERS[0]
  return <Span {...restProps} cls={[props.cls, 'badge', `badge-${mod}`]} />
}
Badge.propTypes = {
  cls: PropTypes.any // todo: classnames.proptypes
}

export const FormGroup = ({
  id,
  label,
  hideLabel,
  labelSmall,
  helpText,
  children,
  ...props
}) => {
  const labelContent = !!(label || labelSmall) && (
    <F>
      {label}
      {!!labelSmall && (
        <F>
          {' '}
          <Small>{labelSmall}</Small>
        </F>
      )}
    </F>
  )

  return (
    <Node {...props} cls="form-group" tag="fieldset">
      {labelContent && (
        <Label htmlFor={id} cls={{ 'sr-only': hideLabel }}>
          {labelContent}{' '}
        </Label>
      )}
      {!!children && <Div>{children} </Div>}
      {helpText && (
        <Small id={`${id}--Help`} cls="form-text text-muted">
          {f.trim(helpText)}{' '}
        </Small>
      )}
    </Node>
  )
}
FormGroup.propTypes = {
  id: PropTypes.string,
  label: PropTypes.node,
  hideLabel: PropTypes.bool,
  labelSmall: PropTypes.node,
  helpText: PropTypes.node,
  children: PropTypes.node
}

export const InputText = props => (
  <ControlledInput {...props}>
    {inputProps => (
      <Node
        autoComplete="off-even-in-chrome"
        {...inputProps}
        tag="input"
        type="text"
        cls={['form-control', inputProps.cls]}
      />
    )}
  </ControlledInput>
)

const FormFieldPropTypes = {
  beforeInput: PropTypes.node,
  afterInput: PropTypes.node,
  helpText: PropTypes.node,
  id: PropTypes.string,
  children: PropTypes.oneOf([null, undefined]),
  label: PropTypes.node.isRequired,
  hideLabel: PropTypes.bool,
  labelSmall: PropTypes.node,
  name: PropTypes.string.isRequired,
  placeholder: PropTypes.string,
  type: PropTypes.string, // enum, already checked at runtime
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
}
export const FormField = ({
  beforeInput,
  afterInput,
  helpText,
  id,
  children,
  label,
  hideLabel,
  labelSmall,
  name,
  placeholder,
  type = 'text',
  value,
  ...inputProps
}) => {
  const supportedTypes = [
    'text',
    'text-static',
    'textarea',
    'number',
    'number-integer',
    'checkbox'
  ]
  if (children) {
    throw new Error('`children` not supported! Use `FormGroup` instead.')
  }

  if (!f.includes(supportedTypes, type)) {
    throw new Error('Unsupported Input Type!')
  }

  let tag = 'input'
  let mainClass = 'form-control'
  if (!id) id = name
  if (!name) name = id

  if (type === 'number-integer') {
    type = 'number'
    inputProps = { min: 1, step: 1, ...inputProps }
  }

  if (type === 'textarea') {
    tag = type
    type = null
    let [minRows, maxRows] = [3, 7]
    inputProps = {
      rows: Math.max(
        minRows,
        Math.min(maxRows, String(value).split('\n').length)
      ),
      ...inputProps
    }
  }

  if (type === 'text-static') {
    tag = 'span'
    mainClass = 'form-control-plaintext'
  }

  const inputNode = (
    <Node
      tag={tag}
      cls={mainClass}
      type={type}
      {...inputProps}
      id={id}
      name={name}
      value={value === null ? '' : value}
      placeholder={placeholder}
      autoComplete="off-even-in-chrome"
      aria-describedby={helpText ? `${id}--Help` : null}
    />
  )

  return (
    <FormGroup
      label={label}
      hideLabel={hideLabel}
      labelSmall={labelSmall}
      helpText={helpText}
    >
      {beforeInput}
      {inputNode}
      {afterInput}
    </FormGroup>
  )
}
FormField.propTypes = FormFieldPropTypes

// like bootstrap docs
// export const FilePicker = ({ id, name }) => (
//   <div className="custom-file">
//     <input type="file" className="custom-file-input" id={id} name={name} />
//     <label className="custom-file-label" htmlFor="customFile">
//       <Icon.Paperclip /> {t('form_filepicker_label')}
//     </label>
//   </div>
// )

export const FilePicker = ({ id, name, label }) => (
  <div>
    <label className="btn btn-sm btn-block btn-flat btn-outline-secondary text-left">
      <Icon.Paperclip /> {label}
      <input id={id} name={name} type="file" style={{ opacity: 0, width: 0 }} />
    </label>
  </div>
)
FilePicker.propTypes = {
  id: PropTypes.string,
  name: PropTypes.string,
  label: PropTypes.node
}
