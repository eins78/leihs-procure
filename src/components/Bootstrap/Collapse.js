import React from 'react'
import PropTypes from 'prop-types'
import Icon from '../Icons'

export class Collapse extends React.Component {
  static defaultProps = { startOpen: false, canToggle: true }
  static propTypes = { id: PropTypes.string.isRequired }

  constructor(props) {
    super()
    this.state = { isOpen: props.canToggle === false || props.startOpen }
  }

  onToggleOpen(event) {
    event.preventDefault()
    if (!this.props.canToggle) return
    this.setState(s => ({ isOpen: !s.isOpen }))
  }

  render({ props: { id, children, canToggle }, state: { isOpen } } = this) {
    const toggleOpen = e => this.onToggleOpen(e)
    const collapsedProps = {
      id: `${id}-content`,
      'aria-labelledby': id ? `${id}-toggle` : false
    }
    return children({
      canToggle,
      toggleOpen,
      isOpen,
      Caret: isOpen ? Icon.CaretDown : Icon.CaretRight,
      collapsedProps,
      togglerProps: {
        onClick: toggleOpen,
        id: `${id}-toggle`,
        'aria-expanded': isOpen ? 'true' : 'false',
        'aria-controls': `${id}-content`
      }
    })
  }
}
