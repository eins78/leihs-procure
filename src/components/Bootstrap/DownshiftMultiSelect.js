import React from 'react'
// import Downshift from 'downshift'

// import ButtonDropdown from 'reactstrap/lib/ButtonDropdown'
// import DropdownToggle from 'reactstrap/lib/DropdownToggle'
// import DropdownMenu from 'reactstrap/lib/DropdownMenu'
// import DropdownItem from 'reactstrap/lib/DropdownItem'

const MultiSelect = ({ options, ...rest }) => {
  return (
    <select multiple {...rest}>
      {options.map(({ label, options }, i) => (
        <optgroup key={i} label={label}>
          {options.map(({ label, ...props }, i) => (
            <option key={i} {...props}>
              {label}
            </option>
          ))}
        </optgroup>
      ))}
    </select>
  )
}

export default MultiSelect

// function titleBySelection(selected, values) {
//   const count = selected.length
//   const valuesCount = values.length
//   if (count === valuesCount) {
//     return 'Alle ausgew\xE4hlt'
//   }
//   if (count === 0) {
//     return 'Keine ausgew\xE4hlt'
//   }
//   if (count > 3) {
//     return `${count} selektiert`
//   }
//   return selected
//     .map(item => values.filter(({ value }) => value === item)[0])
//     .map(item => item && item.label)
//     .join(', ')
// }
