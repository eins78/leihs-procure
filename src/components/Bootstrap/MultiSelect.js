import React from 'react'
import PropTypes from 'prop-types'
import ButtonDropdown from 'reactstrap/lib/ButtonDropdown'
import DropdownToggle from 'reactstrap/lib/DropdownToggle'
import DropdownMenu from 'reactstrap/lib/DropdownMenu'
import DropdownItem from 'reactstrap/lib/DropdownItem'

function titleBySelection(selected, values) {
  const count = selected.length
  if (count === 0) {
    return 'Keine Ausgew\xE4hlt'
  }
  if (count > 3) {
    return `${count} selektiert`
  }
  return selected
    .map(item => values.filter(({ value }) => value === item)[0])
    .map(item => item && item.label)
    .join(', ')
}

class MultiSelect extends React.PureComponent {
  constructor() {
    super()
    this.state = { dropdownOpen: false, selected: [] }
    this.onCheckboxChange = this.onCheckboxChange.bind(this)
    this.onSelectAllChange = this.onSelectAllChange.bind(this)
  }

  componentDidMount() {
    const inititalSelected = this.props.values
      .filter(i => i.selected)
      .map(i => i.value)
    this.setState(() => ({ selected: inititalSelected }))
  }

  toggleDropdown() {
    this.setState(state => ({
      dropdownOpen: !state.dropdownOpen
    }))
  }

  onCheckboxChange(event, value) {
    const isSelected = event.target.checked
    const item = value || event.target.name
    this.setState(
      prev => {
        const selected = prev.selected
          .concat(item)
          .filter(val => (val === item ? isSelected : true))
        return { selected }
      },
      () => this.props.onChange(this.state.selected)
    )
  }

  onSelectAllChange(event) {
    const isSelected = event.target.checked
    this.setState((prev, props) => ({
      selected: !isSelected ? [] : props.values.map(({ id }) => id)
    }))
  }

  render({ props, state } = this) {
    const { values, name, ...restProps } = props
    const allSelected = values.length === state.selected.length

    return (
      <ButtonDropdown
        {...restProps}
        isOpen={this.state.dropdownOpen}
        toggle={() => this.toggleDropdown()}
      >
        <DropdownToggle caret outline>
          {titleBySelection(state.selected, values)}
        </DropdownToggle>
        <DropdownMenu className="multiselect-container">
          <DropdownItem
            className={'multiselect-all ' + (allSelected ? 'active' : '')}
          >
            <a tabIndex="0" className="multiselect-all">
              <label className="checkbox">
                <input
                  type="checkbox"
                  checked={allSelected}
                  onChange={this.onSelectAllChange}
                />
                {'Alle ausw\xE4hlen'}{' '}
              </label>
            </a>
          </DropdownItem>
          {values.map(({ value, label }, ix) => {
            const isSelected = state.selected.indexOf(value) !== -1
            return (
              // <DropdownItem key={ix} className={isSelected ? 'active' : ''}>
              <label
                key={ix}
                className="checkbox"
                data-htmlFor={`${value}-checkbox`}
              >
                <input
                  id={`${value}-checkbox`}
                  type="checkbox"
                  value={value}
                  checked={isSelected}
                  onChange={e => this.onCheckboxChange(e, value)}
                />
                {label}
                {' X '}
              </label>
              // {/* </DropdownItem> */}
            )
          })}
        </DropdownMenu>
      </ButtonDropdown>
    )
  }
}

MultiSelect.defaultProps = {
  name: '',
  values: [],
  onChange: selected => {
    // eslint-disable-next-line no-console
    console.log('MultiSelect', { selected })
  }
}

MultiSelect.propTypes = {
  name: PropTypes.string.isRequired,
  values: PropTypes.arrayOf(
    PropTypes.shape({
      value: PropTypes.string.isRequired,
      label: PropTypes.node.isRequired
    })
  ).isRequired
}

export default MultiSelect
