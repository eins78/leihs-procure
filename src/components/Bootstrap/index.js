export * from './Bootstrap'
export const __Node = null // don't export this into rest of app

export * from './helpers'

export { Collapse } from './Collapse'

export { default as ControlledForm, ControlledInput } from './ControlledForm'

export { default as Select } from './Select'
export { default as ButtonRadio } from './ButtonRadio'
export { default as MultiSelect } from './MultiSelect'
