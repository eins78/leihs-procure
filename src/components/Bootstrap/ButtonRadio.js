import React, { Fragment as F } from 'react'
// import PropTypes from 'prop-types'
import cx from 'classnames'
// import f from 'lodash'
import { Label } from './Bootstrap'
import Icon from '../Icons'

// TODO: responsive when long labels (e.g. btn-group-vertical)
// https://getbootstrap.com/docs/4.0/components/buttons/#checkbox-and-radio-buttons
// TODO: a11y
// https://inclusive-components.design/toggle-button/
// https://www.w3.org/TR/2016/WD-wai-aria-practices-1.1-20160317/examples/radio/radio.html
class ButtonRadio extends React.PureComponent {
  getSelectedValueProp(value, selected) {
    // `value` or `selected` is accepted, depending on consistency
    // with DOM or between components is desired.
    if (value && selected) {
      throw new Error(
        'Props `value` and `selected` were given, please only use 1 of them!'
      )
    }
    return value || selected
  }

  render({ props } = this) {
    const {
      id,
      name,
      value,
      selected,
      onChange,
      options = [],
      withIcons = true,
      ...restProps
    } = props
    const selectedValue = this.getSelectedValueProp(value, selected)
    return (
      <div
        {...restProps}
        className="btn-group btn-group-block"
        role="radiogroup"
      >
        {options.map(({ label, value, ...item }, n) => {
          const inputID = `${id}_radio_${n}`
          const isSelected = value === selectedValue
          return (
            <F key={n}>
              <Label
                cls={cx(
                  'btn btn-block btn-flat btn-outline-secondary m-0 text-left font-weight-bold',
                  {
                    'border-left-0': n !== 0,
                    active: isSelected
                  }
                )}
                htmlFor={inputID}
                tabIndex="-1"
              >
                <input
                  {...item}
                  id={inputID}
                  name={name}
                  value={value}
                  type="radio"
                  aria-checked={isSelected}
                  tabIndex={n === 0 ? 0 : -1}
                  onChange={() => {
                    onChange({
                      target: { name: name, value: value }
                    })
                  }}
                  className="sr-only"
                  autoComplete="off-even-in-chrome"
                />
                {withIcons &&
                  (isSelected ? (
                    <Icon.RadioCheckedOn className="mr-2" />
                  ) : (
                    <Icon.RadioCheckedOff className="mr-2" />
                  ))}
                {label}
              </Label>
            </F>
          )
        })}
      </div>
    )
  }
}

export default ButtonRadio
