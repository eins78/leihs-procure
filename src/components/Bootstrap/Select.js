import React from 'react'
import PropTypes from 'prop-types'
import cx from 'classnames'

import { __Node as Node } from './Bootstrap'

const Select = ({ options, multiple, emptyOption, value, ...props }) => {
  const selectedValue =
    typeof value === 'object'
      ? multiple
        ? value
        : value[0]
      : multiple
        ? [value]
        : value
  if (emptyOption === true) emptyOption = Select.defaultProps.emptyOption
  return (
    <Node
      tag="select"
      {...props}
      className={cx('custom-select', props.className)}
      multiple={multiple}
      value={selectedValue}
    >
      {emptyOption && <option {...emptyOption} />}
      {options.map(({ label, children, ...props }, ix) => (
        <option key={ix} {...props}>
          {children || label}
        </option>
      ))}
    </Node>
  )
}

Select.defaultProps = {
  multiple: false,
  options: [],
  emptyOption: { children: '---', value: '' }
}

Select.propTypes = {
  id: PropTypes.string,
  className: PropTypes.string,
  multiple: PropTypes.bool,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.arrayOf(PropTypes.string),
    PropTypes.arrayOf(PropTypes.number)
  ]),
  options: PropTypes.arrayOf(
    PropTypes.shape({
      value: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
        .isRequired,
      label: PropTypes.node.isRequired
    })
  ),
  emptyOption: PropTypes.oneOfType([
    PropTypes.bool,
    PropTypes.shape({
      children: PropTypes.string.isRequired,
      value: PropTypes.string
    })
  ])
}

export default Select
