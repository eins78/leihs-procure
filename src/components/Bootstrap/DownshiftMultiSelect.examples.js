import React from 'react'

import MultiSelect from './DownshiftMultiSelect'

// # DATA

const exampleMultipleOptgroups = [
  {
    label: 'Group 1',
    options: [
      { value: '1-1', label: 'Option 1.1' },
      { value: '1-2', label: 'Option 1.2', selected: true },
      { value: '1-3', label: 'Option 1.3', selected: true }
    ]
  },
  {
    label: 'Group 2',
    options: [
      { value: '2-1', label: 'Option 2.1' },
      { value: '2-2', label: 'Option 2.2' },
      { value: '2-3', label: 'Option 2.3' }
    ]
  }
]

export const examples = [
  {
    name: 'MultiSelect',
    content: (
      <div>
        <MultiSelect
          id="example-multiple-optgroups"
          options={exampleMultipleOptgroups}
        />
      </div>
    )
  }
]
