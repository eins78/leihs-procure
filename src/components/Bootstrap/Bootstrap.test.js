import React from 'react'
import renderer from 'react-test-renderer'

import { Button } from './Bootstrap'

it('Button: renders correctly', () => {
  const tree = renderer.create(
    <React.Fragment>
      <Button>Button</Button>
      <Button color="primary" title="test title">
        Button
      </Button>
      <Button color="link" title="test title" className="my-class">
        Button
      </Button>
    </React.Fragment>
  )
  expect(tree.toJSON()).toMatchSnapshot()
})
