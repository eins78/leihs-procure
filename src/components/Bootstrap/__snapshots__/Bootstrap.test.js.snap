// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button: renders correctly 1`] = `
Array [
  <button
    className="btn btn-secondary"
    onClick={[Function]}
    type="button"
>
    Button
</button>,
  <button
    className="btn btn-primary"
    onClick={[Function]}
    title="test title"
    type="button"
>
    Button
</button>,
  <button
    className="my-class btn btn-link"
    onClick={[Function]}
    title="test title"
    type="button"
>
    Button
</button>,
]
`;
